File: output/playwright/nested-wires/order-trial/capture.py
"""Own nonpersistent headless Chrome on 5192 (order-trial's owned port). No other server is contacted.
Usage: python3 capture.py baseline | variant-a"""
import json
import subprocess
import sys
from pathlib import Path
ROOT = Path(__file__).resolve().parents[4]
OUT = Path(__file__).resolve().parent
CLI = Path.home() / '.codex/skills/playwright/scripts/playwright_cli.sh'
mode = sys.argv[1] if len(sys.argv) > 1 else 'baseline'
options = {'prefix': str(OUT / mode / mode)}
if mode != 'baseline':
    options['spec'] = json.loads((OUT / mode / 'scene-spec.json').read_text())
def command(*args):
    result = subprocess.run([str(CLI), '-s=m9-ordertrial', *args], cwd=ROOT, text=True, capture_output=True, check=True)
    if '### Error' in result.stdout:
        raise RuntimeError(result.stdout)
    return result.stdout
try:
    command('open', 'http://127.0.0.1:5192/roads-prototype.html?templates')
    command('snapshot')
    source = (OUT / 'capture.mjs').read_text().replace('export async function', 'async function')
    result = command('run-code', 'async () => {\n' + source + '\nreturn capture(page,' + json.dumps(options) + ');\n}')
    report = json.loads(result.split('### Result\n', 1)[1].split('\n###', 1)[0])
    (OUT / mode / 'browser.json').write_text(json.dumps(report, indent=2) + '\n')
    print(f"PASS headless Chrome {report['browserVersion']}; 1920x1440 fit view; {report['nodes']} labels / {report['wires']} wires; layout=1; errors=0; interceptedSpec={report['interceptedSpec']}")
    print(f"MEASURE five loads={[l['roads:navigation-to-ready'] for l in report['loads']]}; median={report['medianMilliseconds']:.3f} ms")
finally:
    command('close')

File: output/playwright/nested-wires/order-trial/capture.mjs
/** Order-trial headless capture body; the Python wrapper owns browser cleanup.
 * options: { prefix, spec? } — when `spec` is present, the scene-spec.json module
 * request is intercepted and fulfilled with that spec (as the vite JSON-module
 * shape the importer expects), so the app renders the reordered variant without
 * any repository file changing. The interception is part of this harness only.
 */
function ensure(condition, message) {
  if (!condition) throw new Error(message);
}
export async function capture(page, options) {
  await page.setViewportSize({ width: 1920, height: 1440 });
  if (options.spec) {
    const body = 'export default ' + JSON.stringify(options.spec) + ';';
    await page.route('**/templates-scene/scene-spec.json**', (route) =>
      route.fulfill({ contentType: 'application/javascript', body }),
    );
  }
  const errors = [];
  page.on('pageerror', (error) => errors.push(error.message));
  const loads = [];
  const url = 'http://127.0.0.1:5192/roads-prototype.html?templates';
  for (let i = 0; i < 5; i += 1) {
    await page.goto(url);
    await page.waitForFunction(
      () => performance.getEntriesByName('roads:navigation-to-ready').length > 0,
    );
    loads.push(
      await page.evaluate(() =>
        Object.fromEntries(
          performance
            .getEntriesByType('measure')
            .filter((e) => e.name.startsWith('roads:'))
            .map((e) => [e.name, e.duration]),
        ),
      ),
    );
  }
  await page.getByRole('button', { name: 'Fit View', exact: true }).click();
  await page.waitForTimeout(350);
  const checkbox = page.getByRole('checkbox', { name: 'Show roads', exact: true });
  await checkbox.check();
  await page.screenshot({ path: `${options.prefix}-roads-on.png` });
  await checkbox.uncheck();
  await page.screenshot({ path: `${options.prefix}-roads-off.png` });
  const labels = await page.locator('[data-node-id] strong').allTextContents();
  const bounds = await page
    .locator('[data-node-id]')
    .evaluateAll((elements) =>
      elements.map((e) => ({
        id: e.getAttribute('data-node-id'),
        ...e.getBoundingClientRect().toJSON(),
      })),
    );
  const report = {
    url,
    interceptedSpec: Boolean(options.spec),
    browserVersion: page.context().browser().version(),
    viewport: page.viewportSize(),
    loads,
    medianMilliseconds: loads
      .map((l) => l['roads:navigation-to-ready'])
      .toSorted((a, b) => a - b)[2],
    nodes: await page.locator('[data-node-id]').count(),
    wires: await page.locator('[data-wire-id]').count(),
    layoutRecalcCount: await page.evaluate(() => window.__layoutRecalcCount),
    labels,
    bounds,
    errors,
    roadsOff: !(await checkbox.isChecked()),
  };
  ensure([report.nodes === 16, report.wires === 29].every(Boolean), 'Rendered graph counts differ');
  ensure(
    labels.length === 16 && labels.every((label) => label.endsWith('.ts')),
    'Real file labels missing',
  );
  ensure(
    bounds.every((b) => b.x >= 0 && b.y >= 0 && b.right <= 1920 && b.bottom <= 1440),
    'Fit view clips nodes',
  );
  ensure(
    [report.layoutRecalcCount === 1, !errors.length].every(Boolean),
    'Layout repeated or browser error',
  );
  return report;
}
